### ISSUE TITLE
Feature Request: Differentiate functions and methods

### ISSUE THREAD

**Issue by Erotemic** [jendrikseipp/vulture#112]
I think it would be useful differentiate functions from methods. 

Consider this use case:

We have a class defined in an external package. It expects a third party to inherit from that class and define a special abstract method. Then when the "run" method of that class is called it does something with that special abstract method

```python

## External package (foo)
class Foo:
    def run(self):
        self.callback()
    
    def callback(self);
         raise NotImplemented('must define')

## Our package
import foo

class Bar(foo.Foo):
    def callback(self):
        print('hello world')

Bar().run()
```

In this case running vulture will report that `callback` is unused. If this were differentiated from functions it would be easier to ignore cases like this using grep.

**Comment by jendrikseipp** [comment]
I'm used to the following terminonlogy: *methods* belong to classes and *functions* are defined outside of classes. Therefore, I'm a little confused by the issue title: I only see methods in your example. Can you clarify, please?

I think in your example, one solution could be to include the external file in the list of files passed to Vulture. Then no method would be reported as unused.

**Comment by Erotemic** [comment]
You have the correct understanding of functions and methods. I was simply giving an example where I wouldn't want to have a method printed. Simply image there is a function there that I want. 

It should be a relatively simple matter to differentiate them. Simply have the ast visitor keep track of its scope and when the current scope is a class, any functions that are encountered can simply be labeled as methods (perhaps even tagged with the class that they belong to as I've done in xdoctest: https://github.com/Erotemic/xdoctest/blob/master/xdoctest/static_analysis.py#L111 Of course the logic would have to be slightly different because In xdoctest I only care about top level definitions). 

The alternative that you've suggested would work, but I think its undesirable because they vulture will report unused items in that external package as well. In fact --- if I understand the code correctly --- it may even cause false negatives because the external package may have functions with conflicting names. 

Another use case for differentiating functions and methods is this: 


```python

class Foo:
    def foo(self);
        pass

def foo():
    pass

# Foo.foo()
# foo()
```

regardless of which of the last two lines you comment vulture will simply report that there is a function named foo that is unused (or perhaps it will report nothing unused if it really only cares about the names, but I'm not sure). Either way, if would be nice if the output specified that there was an unused function `foo` and an unused method `Foo.foo`.

**Comment by jendrikseipp** [comment]
Differentiating between functions and methods is tricky. For our purposes methods could be functions with a "self" parameter that are stored as attributes, not as variables. However, we can't differentiate between function and method executions. Therefore, even if Vulture reports an unused method "foo" and an unused function "foo", once either `Foo.foo()` or `foo()` is uncommented in your example, Vulture would have to declare both the method and the function as used.

**Comment by Dobatymo** [comment]
I also think this would be a very useful feature.

**Comment by erdseb** [comment]
My vote for making the reporting say "unused function/method" rather than "unused function".

**Comment by jendrikseipp** [comment]
> My vote for making the reporting say "unused function/method" rather than "unused function".

I agree that this would be a useful change. Any other opinions?

**Comment by Erotemic** [comment]
Anything that differentiates them is fine.

**Comment by jendrikseipp** [comment]
Thinking about this again, my comment https://github.com/jendrikseipp/vulture/issues/112#issuecomment-395039530 seems to be wrong. Differentiating between functions and methods is beneficial for Vulture and I implemented this in #209.

**Comment by erdseb** [comment]
Awesome! :tada:
